pkg/pgmodel/sql_ingest.go

```diff
@@ -715,44 +764,111 @@ func (h *insertHandler) setSeriesIds(sampleInfos []samplesInfo) (string, error)
 		lastSeenLabel = curr.labels
 	}
 
+	if numSQLFunctionCalls != len(batchSeries) {
+		return "", epoch, fmt.Errorf("unexpected difference in numQueries and batchSeries")
+	}
+
 	br, err := h.conn.SendBatch(context.Background(), batch)
 	if err != nil {
-		return "", err
+		return "", epoch, err
 	}
 	defer br.Close()
 
-	if numSQLFunctionCalls != len(batchSeries) {
-		return "", fmt.Errorf("unexpected difference in numQueries and batchSeries")
+	// BEGIN;
+	_, err = br.Exec()
+	if err != nil {
+		return "", epoch, err
+	}
+
+	var newEpoch int64
+	row := br.QueryRow()
+	err = row.Scan(&newEpoch)
+	if err != nil {
+		return "", epoch, err
+	}
+	if epoch == -1 || newEpoch < epoch {
+		epoch = newEpoch
+	}
+	if h.seriesCacheEpoch == -1 || newEpoch < h.seriesCacheEpoch {
+		h.seriesCacheEpoch = newEpoch
+	}
+
+	// COMMIT;
+	_, err = br.Exec()
+	if err != nil {
+		return "", epoch, err
 	}
 
 	var tableName string
 	for i := 0; i < numSQLFunctionCalls; i++ {
+		// BEGIN;
 		_, err = br.Exec()
 		if err != nil {
-			return "", err
+			return "", epoch, err
 		}
-		row := br.QueryRow()
 
 		var id SeriesID
+		row = br.QueryRow()
 		err = row.Scan(&tableName, &id)
 		if err != nil {
-			return "", err
+			return "", epoch, err
 		}
-		h.seriesCache[batchSeries[i][0].labels.String()] = id
+		seriesString := batchSeries[i][0].labels.String()
+		h.seriesCache[seriesString] = id
 		for _, lsi := range batchSeries[i] {
 			lsi.seriesID = id
 		}
+
+		// COMMIT;
 		_, err = br.Exec()
 		if err != nil {
-			return "", err
+			return "", epoch, err
 		}
 	}
 
-	return tableName, nil
+	return tableName, epoch, nil
 }
 
-func (p *pendingBuffer) addReq(req insertDataRequest) bool {
+func (h *insertHandler) refreshSeriesCache() {
```

Comment:
I think this must be called if the insert fails with a 'PS001' error. That error should cause a refresh and the propagate up. That makes sure that a retry doesn't hit this same error again.

Reply:
I don't see a way to reasonably do without adding a backedge to our gouroutine communication graph, and for something that likely to deadlock I'd rather put it in its own PR, with time to think through all of test cases we should have. I'm also hesitant to delay this PR any longer than it has been; right this bug can cause unrecoverable data corruption, and  fixing that, even if we introduce a brief DoS in an unlikely desynchronization event, seems like enough of an improvement that I don't think we should delay releasing it unless we absolutely have to.

Reply:
(@cevian)